torch_xla/distributed/zero_redundancy_optimizer.py

```diff
@@ -57,11 +69,13 @@ def __init__(
       grad_clipping: bool = True,
       max_norm: Optional[float] = None,
       pin_layout: bool = True,
-      use_grad_acc_hook: bool = False,
       sharding_groups: Optional[Any] = None,
       grad_norm_groups: Optional[Any] = None,
       lazy_init: bool = False,
       coalesce_cc: bool = False,
+      use_grad_acc_hook: bool = False,
+      save_master_weights: bool = False,
+      higher_cc_precision: bool = False,
```

Comment:
why do we need this? If once the optimizer dtype is set, shouldn't all ops in optimizer use the same dtype?

Reply:
it controls the cc ops dtype, 2 reason to have this:
- without this, we see performance drop from tutorial, because it's from bf16/bf16 to fp32/fp32
- Haozheng's gpu compatible branch need this, we make it a config in nxd

[reduce-scatter dtype]/[all-gather dtype]
| use_grad_acc_hook/higher_cc_precision | True      | False     |
| ------------------------------------- | --------- | --------- |
| True                                  | fp32/fp32 | fp32/bf16 |
| False                                 | fp32/fp32 | bf16/bf16 |

Reply:
For production version of this we need to ensure the feature matches the expected use cases. I think there are some gaps. Please refer to use case below. 

Also let me know if I missed anything

Checkpoint save : 
+++++++++++
1. Default use case : Save Master Weights and Optimizer State. Do not save : Forward Weights. 
2. Backward compatibility : Save Forward Weights and Optimizer State. Do not save : Master Weights. 
3. There is no use case for saving all 3 master + forward + optimizer but also no reason to prevent it ? 

DEFAULT BEHAVIOR should be (1). 
WARNINGS : We should print warnings for use case (2) and (3).
  * (2) - warn user that this has accuracy issues if they use it to resume training.
  * (3) - warn user that they are wasting disk space and forward weights will be ignored on load.
NOTE : There may exist use cases where customer want (2) because they want to publish lower precision checkpoint for inference or fine-tuning etc. We recommend that be done using an offline conversion script. A convenience feature maybe to automatically dump an *extra* lower precision checkpoint only on the final checkpoint. Some users continue training even from "completed" training runs with new data so the final checkpoint still needs to have the full precision.

Checkpoint load : 
+++++++++++
1. Default use case : Load master weights and Optimizer. Broadcast that to forward weights and ignore forward weights if they were saved.
2. Backward Compatibility : If master weights does not exist, the run should *FAIL* by default. Then expose an option to force a training to resume from lower precision weights

DEFAULT BEHAVIOR should be (1) and FAIL on (2).
WARNINGS : We should inform the user that continuing training from low precision weights will affect accuracy, especially if there were multiple crashes close in time because the errors introduced by each save/restore will compound itself.  We should print the option needed to override the FAIL in the warning message.

Reply:
agree with these. while these should be done in nxd/nnm level.

for this PR, do we want to use `True` as default for `save_master_weights` and `higher_cc_precision`?

Reply:
Instead of duping code in NNM and NXD, I think the warnings should be in ZERO1. 
IMHO, the default should be to use save_master_weights = True but this means it is a change that must be documented in release notes. 

for `higher_cc_precision` this is a product decision.